### ISSUE TITLE
[`unrelated_type_equality_checks`] Missing report for unrelated extension types

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/linter#4848]
**Describe the issue**
The lint fails to appear when comparing two (unrelated) extension types

**To Reproduce**
```dart
extension type UserID(int id) { }
extension type ProfileID(String id) { }

void main() {
  final a = 1;
  final b = UserID(1);
  final c = ProfileID("Hello");
  print(a == b);  // unrelated
  print(b == c);  // <-- no lint?
  print(a == c);  // unrelated
}
```

This also happens in the subtler case when `ProfileID` is an extension on `int` as well: 

```dart
extension type UserID(int id) { }
extension type ProfileID(int id) { }

void main() {
  final a = 1;
  final b = UserID(1);
  final c = ProfileID(1);
  print(a == b);  // unrelated
  print(b == c);  // <-- no lint?
  print(a == c);  // unrelated
}
```

**Expected behavior**
I would expect comparing `b == c` would show a lint, just like `a == b` and `a == c` does.

**Comment by srawlins** [comment]
Oh ho! It seems this would be an important one to catch. Indeed they are unrelated. The logic might trip up because I think we allow some subtyping or "common supertype" types to be considered "related." If it is important to keep that logic around, we can still not consider "common supertype" extension types to be considered "related."
